compiler/rustc_middle/src/ty/flags.rs

```diff
@@ -301,8 +307,12 @@ impl FlagComputation {
     }
 
     fn add_unevaluated_const(&mut self, ct: ty::Unevaluated<'tcx>) {
-        // TODO
-        self.add_substs(ct.substs_.unwrap());
+        if let Some(substs) = ct.substs_ {
+            self.add_substs(substs);
+        } else {
```

Comment:
>  see rustc-dev-guide

*subtly* implying that I have to write a section for the rustc-dev-guide for us to merge this :p

> When the substs are loaded, they will be the "identity substs" (is this true?) for the constant

not quite, going to write the `rustc-dev-guide` section soon, but e.g.
```rust
fn foo<const N: usize>() -> [u8; N + 1] { todo!() }

fn bar() {
    foo::<3>();
}
```
here the substs for the `N + 1` constant is `3` inside of `bar` while the default substs is the parameter `N`.

So if `substs_` is `Some`, it may either be the default substs - which are all used generic parameters of the parent and, in the future, the bound vars used by the const, e.g. `for<'a> Trait<[u8; foo::<'a>()]>`-  or an instantiation of these params.

> In other words, why do we need HAS_KNOWN_TY_PARAM vs HAS_TY_PARAM?

In other words, where does it cause problems to always assume something has generic parameters if there are unknown anon const substs? 

In wfcheck this would cause us to incorrectly not emit an error here if the default type if the default of a param contained an anon const with unknown substs:
https://github.com/rust-lang/rust/blob/effea9a2a0d501db5722d507690a1a66236933bf/compiler/rustc_typeck/src/check/wfcheck.rs#L856

Most of the other uses of `has_param_types_or_consts` would not cause any bugs as we call `var.subst(tcx, subst)` before that which exposes all default substs. Verifying that isn't trivial though and requires some subtle knowledge about how all of this works, so we should avoid relying on it imo.

For `HAS_FREE_REGIONS` the relevant method is `is_global` which uses the negated type flag, so if we always set `HAS_FREE_REGIONS` and use `var.is_global() == false` to influence user visible behavior this would also be a bug.

I guess that e.g. this would be an issue https://github.com/rust-lang/rust/blob/effea9a2a0d501db5722d507690a1a66236933bf/compiler/rustc_typeck/src/impl_wf_check/min_specialization.rs#L366